source/Halibut.Tests/Transport/Streams/TimeoutStreamWrapperFixture.cs

```diff
@@ -0,0 +1,369 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Halibut.Tests.Support;
+using Halibut.Tests.Support.TestAttributes;
+using Halibut.Transport.Streams;
+using NUnit.Framework;
+
+namespace Halibut.Tests.Transport.Streams
+{
+    [TestTimeout(timeoutInSeconds: 20)]
+    public class TimeoutStreamWrapperFixture : BaseTest
+    {
+        [Test]
+        public async Task ReadShouldPassThrough()
+        {
+            string? readData = null;
+            
+            var (disposables, sut, performListenerWrite) = await BuildTcpClientAndTcpListener(CancellationToken);
+
+            using (disposables)
+            {
+                await performListenerWrite("Test");
+
+                var buffer = new byte[19];
+                var readBytes = sut.Read(buffer, 0, 19);
+                if (readBytes > 0)
+                {
+                    readData = Encoding.UTF8.GetString(buffer, 0, readBytes);
+                }
+
+                Assert.AreEqual("Test", readData);
+            }
+        }
+        
+        [Test]
+        public async Task ReadAsyncShouldPassThrough()
+        {
+            string? readData = null;
+            
+            var (disposables, sut, performListenerWrite) = await BuildTcpClientAndTcpListener(CancellationToken);
+
+            using (disposables)
+            {
+                await performListenerWrite("Test");
+
+                var buffer = new byte[19];
+                var readBytes = await sut.ReadAsync(buffer, 0, 19, CancellationToken);
+                if (readBytes > 0)
+                {
+                    readData = Encoding.UTF8.GetString(buffer, 0, readBytes);
+                }
+
+                Assert.AreEqual("Test", readData);
+            }
+        }
+
+        [Test]
+        public async Task ReadAsyncShouldCancel()
+        {
+            using (var readTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(5)))
+            {
+                var (disposables, sut, _) = await BuildTcpClientAndTcpListener(
+                    CancellationToken,
+                    onListenerRead: async _ => await Task.Delay(TimeSpan.FromDays(1), CancellationToken));
```

Comment:
This lambda gets called straight after we "Read". And will effectively wait until the test finishes (because it waits a day, and uses the test's CancellationToken).

This stops the 'read' loop.

I wonder if it's worth a comment maybe? It wasn't obvious to me when I first looked at it (might be a reflection on me maybe 😁 )

Reply:
We could even make a method that does this, and call it `DelayForADayToDelayReading` or something like that (and use it in the other tests as well). Just a thought.

Reply:
Good spot, this was actually confusing as it was used in the read tests when it was not needed. It's only relevant for the write tests and is an attempt to delay one of the write attempts for timeout and cancellation testing. Cleaned up and `DelayForeverToTryAndDelayWriting` added for clarity of the intent of the delay